oracles/src/repositories/oracle_repository.py

```diff
@@ -257,6 +263,179 @@ async def mark_function_call_as_done(self, function_call: FunctionCall):
+            )
+            for i in range(self.last_kb_query_count, kb_query_count):
+                callback_id = await self.oracle_contract.functions.kbQueryCallbackIds(
+                    i
+                ).call()
+                is_processed = await self.oracle_contract.functions.isKbQueryProcessed(
+                    i
+                ).call()
+                request = await self.oracle_contract.functions.kbQueries(i).call()
+                cid = request[0]
+                query = request[1]
+                num_documents = request[2]
+                index_cid = await self.oracle_contract.functions.kbIndexes(cid).call()
+                self.indexed_kb_queries.append(
+                    KnowledgeBaseQuery(
+                        id=i,
+                        callback_id=callback_id,
+                        is_processed=is_processed,
+                        cid=cid,
+                        index_cid=index_cid,
+                        query=query,
+                        num_documents=num_documents,
+                    )
+                )
+            self.last_kb_query_count = kb_query_count
+
+    async def get_unanswered_kb_queries(self) -> List[KnowledgeBaseQuery]:
+        await self._index_new_kb_queries()
+        unanswered_kb_queries = [
+            kb_query
+            for kb_query in self.indexed_kb_queries
+            if not kb_query.is_processed
+        ]
+        return unanswered_kb_queries
+
+    async def send_kb_query_response(
+        self,
+        request: KnowledgeBaseQuery,
+        documents: List[str],
+        error_message: str = "",
+    ) -> bool:
+        nonce = await self.web3_client.eth.get_transaction_count(self.account.address)
+        tx_data = {
+            "from": self.account.address,
+            "nonce": nonce,
+            # TODO: pick gas amount in a better way
+            # "gas": 1000000,
+            "maxFeePerGas": self.web3_client.to_wei("2", "gwei"),
+            "maxPriorityFeePerGas": self.web3_client.to_wei("1", "gwei"),
+        }
+        if chain_id := settings.CHAIN_ID:
+            tx_data["chainId"] = int(chain_id)
+        try:
+            tx = await self.oracle_contract.functions.addKnowledgeBaseQueryResponse(
+                request.id, request.callback_id, documents, error_message
+            ).build_transaction(tx_data)
+        except Exception as e:
+            request.is_processed = True
+            request.transaction_receipt = {"error": str(e)}
+            await self.mark_kb_query_as_done(request)
+            return False
+        tx_receipt = await self._sign_and_send_tx(tx)
+        request.transaction_receipt = tx_receipt
+        request.is_processed = bool(tx_receipt.get("status"))
+        return bool(tx_receipt.get("status"))
+
+    async def mark_kb_query_as_done(self, query: KnowledgeBaseQuery):
+        nonce = await self.web3_client.eth.get_transaction_count(self.account.address)
+        tx_data = {
```

Comment:
Also we are duplicating this tx_data creation everywhere, just a sidenote here but the sooner we eliminate it the easier I guess :D but not urgent

Reply:
I agree, the whole codebase needs proper vacuuming.